src/liballoc/vec.rs

```diff
@@ -603,6 +603,92 @@ impl<T> Vec<T> {
+    /// The target type must have the same size and alignment as the source type.
+    /// This API doesn't transmute any values of T to U, because it makes sure
+    /// to empty the vector before any unsafe operations.
+    ///
+    /// # Example
+    ///
+    /// This API is useful especially when using `Vec` as a
+    /// temporary storage for data with short lifetimes.
+    /// By recycling the allocation, the `Vec` is able to safely
+    /// outlive the lifetime of the type that was stored in it.
+    /// ```
+    /// # #![feature(recycle_vec)]
+    /// # use std::error::Error;
+    /// #
+    /// # struct Stream(bool);
+    /// #
+    /// # impl Stream {
+    /// #     fn new() -> Self {
+    /// #         Stream(false)
+    /// #     }
+    /// #
+    /// #     fn next(&mut self) -> Option<&[u8]> {
+    /// #         if self.0 {
+    /// #             None
+    /// #         } else {
+    /// #             self.0 = true;
+    /// #             Some(&b"foo"[..])
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # fn process(input: &[Object<'_>]) -> Result<(), Box<dyn Error>> {
+    /// #     for obj in input {
+    /// #         let _ = obj.reference;
+    /// #     }
+    /// #     Ok(())
+    /// # }
+    /// #
+    /// # struct Object<'a> {
+    /// #     reference: &'a [u8],
+    /// # }
+    /// #
+    /// # fn deserialize<'a>(input: &'a [u8], output: &mut Vec<Object<'a>>) -> Result<(), Box<dyn Error>> {
+    /// #     output.push(Object { reference: input });
+    /// #     Ok(())
+    /// # }
+    /// #
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// # let mut stream = Stream::new();
+    /// let mut objects: Vec<Object<'static>> = Vec::new();    // Any lifetime goes here
+    ///
+    /// while let Some(byte_chunk) = stream.next() {           // `byte_chunk` lifetime starts
+    ///     let mut temp: Vec<Object<'_>> = objects.recycle(); // `temp` lifetime starts
+    ///
+    ///     // Zero-copy parsing; deserialized `Object`s have references to `byte_chunk`
+    ///     deserialize(byte_chunk, &mut temp)?;
+    ///     process(&temp)?;
+    ///
+    ///     objects = temp.recycle();                          // `temp` lifetime ends
+    /// }                                                      // `byte_chunk` lifetime ends
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    /// Panics if the size or alignment of the source and target types don't match.
+    ///
+    /// # Note about stabilization
+    /// The size and alignment contract is enforceable at compile-time,
+    /// so we will want to wait until compile-time asserts become stable and
+    /// modify this API to cause a compile error instead of panicking
+    /// before stabilizing it.
+    #[unstable(feature = "recycle_vec", reason = "new API", issue = "0")]
+    pub fn recycle<U>(mut self) -> Vec<U> {
+        self.truncate(0);
+        // TODO make these const asserts once it becomes possible
```

Comment:
Is there an actual plan for making this possible? Even with compile time assertions, you wouldn't actually be able to reference `T` and `U` inside e.g. a `const` item defined in this body.

Reply:
That is a question that needs to be resolved. It is mentioned as an unresolved question in the RFC. The problem is that changing to compile-time error is a breaking change, so whatever gets decided, we must commit to before stabilizing.

I assumed that once the const functions support conditionals, creating compile time assertion for this case would be possible, but I had a hidden assumption that if there is a generic const function, I could pass it generic types that are in scope. Is the latter not the case? I wonder if there explicit plans about supporting or not supporting that, or is it just more about how it currently is?

Reply:
That's not the case and there are no current plans to change that so don't expect it to change within the next 1.5 years at least.

Reply:
All right, I removed the `FIXME`. I guess it seems good enough just to panic, although I'd love to have a compile-time guarantee, and I think I'm not alone. I wonder if there is any other avenue of attack that would accomplish this.